### PR TITLE
PSA implementation of ed25519 image verification and x25519 random key encryption/decryption

### PR DESCRIPTION
**Pull request by de-nordic**
The PSA encryption is quite complete.
The mbedTLS configuration is rather dodgy.
There are few requirements from upstream and mcuboot upmerge would be nice.

The TODO for reviewers:
 - check licenses on [noups]

### CODE REVIEW

File: boot/zephyr/Kconfig [hunk 1]
```diff
@@ -60,6 +70,50 @@ config NRF_CC310_BL
 	bool
 	default n
 
+if BOOT_USE_PSA_CRYPTO
+config BOOT_ED25519_PSA_DEPENDENCIES
+	bool
+	default n
+	select PSA_WANT_ALG_SHA_256
```
Comment (nvlsianpu) [inline boot/zephyr/Kconfig:77]: PSA_WANT_ALG_SHA_256 is stranger there - won't be there once we will introduce corrected signature right?

Comment (de-nordic) [reply boot/zephyr/Kconfig:77]: Yes, that would be with https://github.com/nrfconnect/sdk-mcuboot/pull/325 and https://github.com/mcu-tools/mcuboot/pull/2029

File: boot/bootutil/src/ed25519_psa.c [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    psa_set_key_usage_flags(&key_attr, PSA_KEY_USAGE_VERIFY_MESSAGE);
+    psa_set_key_algorithm(&key_attr, PSA_ALG_PURE_EDDSA);
+
+    status = psa_import_key(&key_attr, public_key, EDDSA_KEY_LENGTH, &kid);
+    if (status != PSA_SUCCESS) {
+        BOOT_LOG_ERR("ED25519 key import failed %d", status);
+        return 0;
+    }
+
+    status = psa_verify_message(kid, PSA_ALG_PURE_EDDSA, message, message_len,
```
Comment (nvlsianpu) [inline boot/bootutil/src/ed25519_psa.c:52]: Is this the place where psa_verify_hash() will be used instead for full sha512 based signature check?

Comment (de-nordic) [reply boot/bootutil/src/ed25519_psa.c:52]: Yes. Exactly. In the following PR https://github.com/nrfconnect/sdk-mcuboot/pull/325/commits/33a7ec498aebb2de8d87f0a5c9aacb76411879b9 there is already that part done.

The difference is that instead of the sha256, which is the `messge` here, we will get sha512, and therefore we do not rely on the internal sha512 calculation of `psa_verify_message`, but rather just verify the hash.

**Review by nvlsianpu — APPROVED** [review]

File: boot/bootutil/src/image_ed25519.c [hunk 3]
```diff
@@ -9,6 +9,11 @@
 
 #include "mcuboot_config/mcuboot_config.h"
 
+#if defined(CONFIG_NRF_SECURITY)
+/* We are not really using the MBEDTLS but need the ASN.1 parsing funcitons */
```
Comment (valeriosetti) [inline boot/bootutil/src/image_ed25519.c:13]: Minor detail, but I don't agree much with this comment. You are not using Mbed TLS for crypto, but you are using its ASN1 parse capabilities. So I would rephrase it as:

/* Enable ASN1 parse functions from Mbed TLS */

Wdyt?

**Review by valeriosetti — COMMENTED** [review]
I just had a quick look to the last 2 commits of this PR and I added some minor comments. I'll take a deeper look ASAP

File: boot/bootutil/src/image_ed25519.c (hunk 3 above)
Comment (de-nordic) [reply boot/bootutil/src/image_ed25519.c:13]: Yeah, I can replace that. btw:https://github.com/nrfconnect/sdk-mcuboot/pull/326

File: boot/bootutil/src/encrypted_psa.c [hunk 4]
```diff
@@ -0,0 +1,461 @@
...
+    /* The psa_cipher_decrypt needs initialization vector of proper length at
+     * the beginning of the input buffer.
+     */
+    uint8_t iv_and_key[PSA_CIPHER_IV_LENGTH(PSA_KEY_TYPE_AES, PSA_ALG_CTR) +
+                       BOOTUTIL_CRYPTO_AES_CTR_KEY_SIZE];
+
+    psa_ret = psa_crypto_init();
+    if (psa_ret != PSA_SUCCESS) {
+        BOOT_LOG_ERR("AES crypto init failed %d", psa_ret);
+        return -1;
```
Comment (valeriosetti) [inline boot/bootutil/src/encrypted_psa.c:148]: I think that you might add `memset(private_key, 0, sizeof(private_key));` here as well just in case `psa_crypto_init` fails.

File: boot/bootutil/src/encrypted_psa.c [hunk 5]
```diff
@@ -0,0 +1,461 @@
...
+        BOOT_LOG_ERR("Shared key import failed %d", psa_ret);
+        return -1;
+    }
+
+    key_do_alg = PSA_ALG_KEY_AGREEMENT(PSA_ALG_ECDH, PSA_ALG_HKDF(PSA_ALG_SHA_256));
+
+    psa_ret = psa_key_derivation_setup(&key_do, key_do_alg);
+    if (psa_ret != PSA_SUCCESS) {
+        BOOT_LOG_ERR("Key derivation setup failed %d", psa_ret);
+        return -1;
```
Comment (valeriosetti) [inline boot/bootutil/src/encrypted_psa.c:168]: Here you might add `psa_destroy_key(kid);` just in case `psa_key_derivation_setup` fails.
From a practical point of view this might not matter much because the device won't boot in this case and the key is volatile (so it will be lost on power cycle), but it might be nice for cleanliness.

File: boot/bootutil/src/encrypted.c [hunk 6]
```diff
@@ -415,7 +387,7 @@ static int fake_rng(void *p_rng, unsigned char *output, size_t len)
  * @param enckey An AES-128 or AES-256 key sized buffer to store to plain key.
  */
 int
-boot_enc_decrypt(const uint8_t *buf, uint8_t *enckey)
+boot_decrypt_key(const uint8_t *buf, uint8_t *enckey)
```
Comment (valeriosetti) [inline boot/bootutil/src/encrypted.c:390]: Why not moving these functions out of `encrypted.c` for better code organization? In this case you would have 2 source files implementing these functions, i.e. `encrypted_psa.c` and the new one, without a huge `#if !defined(CONFIG_BOOT_ED25519_PSA)` in `encrypted.c`...

**Review by valeriosetti — COMMENTED** [review]
I completed the 2nd pass I mentioned this morning and I left some new comments.

File: boot/bootutil/src/encrypted.c (hunk 6 above)
Comment (de-nordic) [reply boot/bootutil/src/encrypted.c:390]: I do not want to do that within this PR, that is something I would like to do, but that has to be done upstream.
Because we are trying to do things with [nrf noups] till we are able to push the changes upstream, I do not want to divert from the upstream code too much.

File: boot/bootutil/src/encrypted_psa.c (hunk 4 above)
Comment (de-nordic) [reply boot/bootutil/src/encrypted_psa.c:148]: I will re-order operations so that the private key is not loaded before crypto init happens.

File: boot/bootutil/zephyr/CMakeLists.txt [hunk 7]
```diff
@@ -34,7 +34,13 @@ target_include_directories(MCUBOOT_BOOTUTIL INTERFACE
 )
 endif()
 
-if(CONFIG_BOOT_USE_MBEDTLS)
+if(CONFIG_BOOT_USE_PSA_CRYPTO)
+target_include_directories(MCUBOOT_BOOTUTIL INTERFACE
+   ${ZEPHYR_MBEDTLS_MODULE_DIR}/include
+)
```
Comment (nordicjm) [inline boot/bootutil/zephyr/CMakeLists.txt:40]: 2 space indent for cmake. Is this going to be submitted upstream?

File: boot/zephyr/CMakeLists.txt [hunk 8]
```diff
@@ -51,6 +51,12 @@ if(EXISTS targets/${BOARD}.h)
   zephyr_library_compile_definitions(MCUBOOT_TARGET_CONFIG="${BOARD}.h")
 endif()
 
+if(DEFINED CONFIG_MBEDTLS)
+zephyr_library_include_directories(
+  ${ZEPHYR_MBEDTLS_MODULE_DIR}/include
+)
```
Comment (nordicjm) [inline boot/zephyr/CMakeLists.txt:57]: as above

File: boot/bootutil/zephyr/CMakeLists.txt (hunk 7 above)
Comment (nordicjm) [reply boot/bootutil/zephyr/CMakeLists.txt:40]: Having a look below, wouldn't be keeping the existing one and adding a new one for psa crypto work better since you're not changing a piece of the upstream code?

File: boot/zephyr/Kconfig [hunk 9]
```diff
@@ -11,6 +11,19 @@ comment "MCUboot-specific configuration options"
...
+config NRF_SECURITY
+	bool
+	default y
+	select NRF_OBERON
+	select BOOT_USE_PSA_CRYPTO
+
+# NRF_SECURITY enforces PSA crypt
+config MBEDTLS
+	bool
+	default n if NRF_SECURITY
```
Comment (nordicjm) [inline boot/zephyr/Kconfig:24]: `NRF_SECURITY` already exists, it should not be getting defined here again, likewise with MBEDTLS. Default should go in a Kconfig.defconfig file

File: boot/zephyr/Kconfig [hunk 10]
```diff
@@ -27,6 +40,15 @@ config BOOT_USE_MBEDTLS
 	help
 	  Use mbedTLS for crypto primitives.
 
+config BOOT_USE_PSA_CRYPTO
+	bool
+	# Hidden option
```
Comment (nordicjm) [inline boot/zephyr/Kconfig:45]:
```suggestion
```

There should not be a comment for this, put hidden option in the help text only

File: boot/zephyr/Kconfig [hunk 11]
```diff
@@ -60,6 +82,58 @@ config NRF_CC310_BL
...
+config MBEDTLS_HEAP_SIZE
+	int
+	default 2048
+	help
+	  The PSA internals need to be able to allocate memory for operation
+	  and it uses mbedTLS heap for that.
+
+config MBEDTLS_ENABLE_HEAP
+	bool
+	default y
```
Comment (nordicjm) [inline boot/zephyr/Kconfig:133]: same with these

File: boot/zephyr/Kconfig [hunk 12]
```diff
@@ -70,13 +144,48 @@ config SINGLE_APPLICATION_SLOT
...
+	  Hidden option to set by configurations that allow SHA384
+
+choice BOOT_IMG_HASH_ALG
+	prompt "Selected image hash algorithm"
+	help
+	  Hash algorithm used for image verification. Selection
+	  here may be limited by other configurations, like for
+	  example selected cryptographic signature.
+	default BOOT_IMG_HASH_ALG_SHA256 if BOOT_IMG_HASH_ALG_SHA256_ALLOW
+	default BOOT_IMG_HASH_ALG_SHA384 if BOOT_IMG_HASH_ALG_SHA384_ALLOW
```
Comment (nordicjm) [inline boot/zephyr/Kconfig:164]: help text always goes last

File: boot/zephyr/Kconfig [hunk 13]
```diff
@@ -117,18 +226,28 @@ endif
...
 	bool "Edwards curve digital signatures using ed25519"
 	select BOOT_ENCRYPTION_SUPPORT
+	select BOOT_IMG_HASH_ALG_SHA256_ALLOW
 
 if BOOT_SIGNATURE_TYPE_ED25519
 choice BOOT_ED25519_IMPLEMENTATION
 	prompt "Ecdsa implementation"
+	default BOOT_ED25519_PSA if NRF_SECURITY
 	default BOOT_ED25519_TINYCRYPT
 config BOOT_ED25519_TINYCRYPT
```
Comment (nordicjm) [inline boot/zephyr/Kconfig:236]: newline

File: boot/zephyr/Kconfig [hunk 14]
```diff
@@ -117,18 +226,28 @@ endif
...
 	bool "Use tinycrypt"
 	select BOOT_USE_TINYCRYPT
+	depends on !NRF_SECURITY
 config BOOT_ED25519_MBEDTLS
 	bool "Use mbedTLS"
 	select BOOT_USE_MBEDTLS
 	select MBEDTLS
+	depends on !NRF_SECURITY
+config BOOT_ED25519_PSA
```
Comment (nordicjm) [inline boot/zephyr/Kconfig:245]: newline

**Review by nordicjm — CHANGES_REQUESTED** [review]

File: boot/bootutil/include/bootutil/crypto/aes_ctr.h [hunk 15]
```diff
@@ -38,12 +38,47 @@
     #define BOOTUTIL_CRYPTO_AES_CTR_BLOCK_SIZE TC_AES_BLOCK_SIZE
 #endif /* MCUBOOT_USE_TINYCRYPT */
 
+
+#if defined(MCUBOOT_USE_PSA_CRYPTO)
+#include <psa/crypto.h>
+    #include "bootutil/enc_key_public.h"
+    /* To fix: why is it always named BOOTUTIL_CRYPTO_AES_CTR_KEY_SIZE */
+    #define BOOTUTIL_CRYPTO_AES_CTR_KEY_SIZE BOOT_ENC_KEY_SIZE
+    #define BOOTUTIL_CRYPTO_AES_CTR_BLOCK_SIZE (16)
```
Comment (nordicjm) [inline boot/bootutil/include/bootutil/crypto/aes_ctr.h:47]: remove indent

File: boot/bootutil/src/image_ed25519.c [hunk 16]
```diff
@@ -73,7 +78,8 @@ bootutil_verify_sig(uint8_t *hash, uint32_t hlen, uint8_t *sig, size_t slen,
     uint8_t *pubkey;
     uint8_t *end;
 
-    if (hlen != 32 || slen != 64) {
+    /* Fixme: Fix needed here, as ed25519 should only use sha512 */
+    if (!(hlen == 64 || hlen == 32) || slen != 64) {
```
Comment (nordicjm) [inline boot/bootutil/src/image_ed25519.c:82]: should be using defines for these sizes

File: boot/bootutil/zephyr/CMakeLists.txt (hunk 7 above)
Comment (de-nordic) [reply boot/bootutil/zephyr/CMakeLists.txt:40]: The general idea is to, at some point, bring this upstream.
I am keeping the PSA code separate because the change is significant and the code is already badly ifdefed in the verification/encryption calls.

File: boot/bootutil/include/bootutil/crypto/aes_ctr.h (hunk 15 above)
Comment (de-nordic) [reply boot/bootutil/include/bootutil/crypto/aes_ctr.h:47]: Why? The entire file is done this way.

Comment (nordicjm) [reply boot/bootutil/include/bootutil/crypto/aes_ctr.h]: We should fix that upstream, it looks silly

**Comment by nvlsianpu** [comment]
@nordicjm Please revisit.

**Comment by nvlsianpu** [comment]
@nordicjm Please revisit.

File: boot/bootutil/zephyr/CMakeLists.txt (hunk 7 above)
Comment (nordicjm) [reply boot/bootutil/zephyr/CMakeLists.txt:40]: Not resolved

File: boot/zephyr/CMakeLists.txt (hunk 8 above)
Comment (nordicjm) [reply boot/zephyr/CMakeLists.txt:57]: Not resolved

File: boot/zephyr/CMakeLists.txt [hunk 17]
```diff
@@ -230,11 +240,19 @@ elseif(CONFIG_BOOT_SIGNATURE_TYPE_ED25519 OR CONFIG_BOOT_ENCRYPT_X25519)
...
+    )
+  else()
+    zephyr_library_sources(
+      ${MBEDTLS_ASN1_DIR}/src/asn1parse.c
+      ${BOOT_DIR}/bootutil/src/ed25519_psa.c
+    )
+  endif()
 endif()
 
+if(NOT CONFIG_BOOT_ED25519_PSA)
```
Comment (nordicjm) [inline boot/zephyr/CMakeLists.txt:255]: 2 space indent below

File: boot/zephyr/Kconfig (hunk 10 above)
Comment (nordicjm) [reply boot/zephyr/Kconfig:32]: Not resolved

**Comment by nordicjm** [comment]
> @nordicjm Please revisit.

None of the comments have been resolved and issues remain

File: boot/zephyr/Kconfig [hunk 18]
```diff
@@ -27,6 +27,15 @@ config BOOT_USE_MBEDTLS
 	help
 	  Use mbedTLS for crypto primitives.
 
+config BOOT_USE_PSA_CRYPTO
+	bool
+	default y if NRF_SECURITY
+	default n
```
Comment (nordicjm) [inline boot/zephyr/Kconfig:33]:
```suggestion
```
`default n` not needed

File: boot/zephyr/Kconfig [hunk 19]
```diff
@@ -60,6 +69,58 @@ config NRF_CC310_BL
...
+config BOOT_PSA_IMG_HASH_ALG_SHA256_DEPENDENCIES
+	bool
+	default y if BOOT_IMG_HASH_ALG_SHA256
+	select PSA_WANT_ALG_SHA_256
+	help
+	  Dependencies for hashing with SHA256
+
+config BOOT_ED25519_PSA_DEPENDENCIES
+	bool
+	default n
```
Comment (nordicjm) [inline boot/zephyr/Kconfig:82]: as above

File: boot/zephyr/Kconfig [hunk 20]
```diff
@@ -60,6 +69,58 @@ config NRF_CC310_BL
...
+	select PSA_WANT_ALG_PURE_EDDSA
+	select PSA_WANT_ECC_TWISTED_EDWARDS_255
+	select PSA_WANT_KEY_TYPE_ECC_KEY_PAIR_IMPORT
+	help
+	  Dependencies for ed25519 signature
+
+if BOOT_ENCRYPT_IMAGE
+config BOOT_X25519_PSA_DEPENDENCIES
+	bool
+	default n
```
Comment (nordicjm) [inline boot/zephyr/Kconfig:94]: as above

File: boot/zephyr/Kconfig [hunk 21]
```diff
@@ -60,6 +69,58 @@ config NRF_CC310_BL
...
+	  encryption. The PSA_WANT_ALG_CTR and PSA_WANT_KEY_TYPE_AES
+	  enable Counter based block cipher and AES key, and algorithm support,
+	  to use with it; the others are used for shared key decryption
+	  and derivation.
+endif
+
+if MBEDTLS_ENABLE_HEAP
+
+config MBEDTLS_HEAP_SIZE
+	int
```
Comment (nordicjm) [inline boot/zephyr/Kconfig:114]:
```suggestion
```

This is an existing symbol already https://github.com/zephyrproject-rtos/zephyr/blob/main/modules/mbedtls/Kconfig#L191

**Review by nordicjm — COMMENTED** [review]
Minor Kconfig nit.

**Review by nordicjm — APPROVED** [review]
